src/rate_limiter/src/lib.rs

```diff
@@ -130,45 +113,30 @@ impl TokenBucket {
         if size == 0 || complete_refill_time_ms == 0 {
             return None;
         }
-        // Formula for computing current refill amount:
-        // refill_token_count = (delta_time * size) / (complete_refill_time_ms * 1_000_000)
-        // In order to avoid overflows, simplify the fractions by computing greatest common divisor.
-
-        let complete_refill_time_ns = complete_refill_time_ms * NANOSEC_IN_ONE_MILLISEC;
-        // Get the greatest common factor between `size` and `complete_refill_time_ns`.
-        let common_factor = gcd(size, complete_refill_time_ns);
-        // The division will be exact since `common_factor` is a factor of `size`.
-        let processed_capacity: u64 = size / common_factor;
-        // The division will be exact since `common_factor` is a factor of
-        // `complete_refill_time_ns`.
-        let processed_refill_time: u64 = complete_refill_time_ns / common_factor;
 
         Some(TokenBucket {
             size,
             one_time_burst,
             initial_one_time_burst: one_time_burst,
             refill_time: complete_refill_time_ms,
+            refill_one_ns: (1_000_000 * complete_refill_time_ms) / size,
```

Comment:
Technically, this just moves the integer division problem from "bucket with very small size" to "bucket with very large size". If `size > 1_000_000 * complete_refill_time_ms`, then `refill_one_ns` will be zero and in `auto_replenish` we perform a division by zero (so instead of hanging we crash).

Reply:
Good catch!!
I think the solution here is to somehow adjust the 100 msec period over which we (gradually) replenish the bucket, e.g. Find the minimum period for which the new tokens would be > 0

Reply:
As @acatangiu noted to overflow here requires `complete_refill_time_ms >= 1.8446744e+13` which is `~213503` days or `~584` years. I think crashing in this case is not unreasonable.

Reply:
```
// A TokenBucket with 1GB/sec allowance
TokenBucket::new(2_000_000_000, 0, 1000)
```

would yield

```
refill_one_ns: (1_000_000 * 1000) / 2_000_000_000
```

which is 0. Am I missing something?

Reply:
> I think the solution here is to somehow adjust the 100 msec period over which we (gradually) replenish the bucket, e.g. Find the minimum period for which the new tokens would be > 0

Yes, that sounds right. I _think_ if we go back to the old implementation, but only execute `last_update = Instant::now()` if more than 0 tokens were replenished would have the effect you propose.

Reply:
When `complete_refill_time_ms = 1000` and `size = 2_000_000_000` we are restoring 1 token every 0.5ns. Do any I/O operations in Firecracker can occur near this frequency?

I don't think most rate limiters consider any case near this frequency, for example https://docs.rs/governor/0.5.1/governor/struct.Quota.html#method.per_second at max allows replacing 4 tokens every ns.

Reply:
> Yes, that sounds right. I think if we go back to the old implementation, but only execute last_update = Instant::now() if more than 0 tokens were replenished would have the effect you propose.

I think this is the best solution here, I've changed it to this.

Reply:
This solution still drops tokens, e.g.:
```rust
#[test]
fn test_token_bucket_auto_replenish() {
    const SIZE: u64 = 1000;
    const TIME: u64 = 1000;
    let time = Duration::from_millis(TIME);

    let mut tb = TokenBucket::new(SIZE, 0, TIME).unwrap();
    tb.reduce(SIZE);
    assert_eq!(tb.budget(), 0);

    let now = Instant::now();
    while now.elapsed() < time {
        tb.auto_replenish();
    }
    assert_eq!(tb.budget(), SIZE - 1);
}
```
But I think this is reasonable.

Reply:
It doesn't drop tokens. I think that if you had put a `tb.auto_replenish()` just after the while loop you would reach `SIZE`. However, see comment below.

Reply:
I did test this, it stills drops tokens

Reply:
weird, with the new test you can't see that, but it's probably some weird timing issue.

Reply:
Ok that makes total sense. I don't think we have a way out of this.